### ISSUE TITLE
Unexpected predictions

### ISSUE THREAD

**Issue by Blair-Young** [scikit-learn-contrib/hdbscan#198]
Hi, 
This is a fantastic tool however, I've come across some unexpected behavior that I'm not sure how to explain. 

### Background
I am currently working with clustering geo-spatial coordinates. I am following the advice in the FAQ by converting the coordinates to Rad before I initiate the clustering. Upon clustering the majority of the predictions make sense (approx. 60 clusters for 446,000 points) .

#### Problem 1. 
There is a particular area of interest which has a high density/concentration of points. I would have expected these clusters to form into a single cluster entity. However, there are multiple clusters within roughly the same area (approx. 1000+ points within each cluster). 

A possible reason for this behavior could be due a large number of samples having exactly the same coordinates. Could these areas of high density be so dense that the model forms separate clusters even though they are only meters apart in real life? (In relative terms the coordinates are distributed across the entire of the UK.)

#### Problem 2.
I then used this trained model to try and predict cluster labels for new points. A significant portion of the predictions make sense. However, the few that are unexpected are in relation to the clusters depicted in **problem 1**. I expect a sample that is close/between (within approx 10 meters) the highly dense clusters to be labeled as one of those clusters, however, it receives a -1 for noise.

Is this due to again the initial clusters appearing so dense, that by proxy- the area around it appears to have very low density- thus being labeled as noise?

[Edit]
#### Problem 3
When predicting a **single** sample's cluster allocation it appears to always return noise.
```
test_sample = np.array([[x_coord, y_coord]])
hdbscan.approximate_predict(HDBSCAN_model, test_sample)
```
Returns- `(array([-1]), array([0.]))`

This seems a bit strange as when predicting a batch (including this exact coordinate) the prediction associates the coordinate with cluster 22, which appears visually correct.

Perhaps I should just tweak my hyperparameter, but I thought it would be interesting to ask anyway.

```hdbscan.HDBSCAN(min_cluster_size=100,gen_min_span_tree=True, metric='haversine', prediction_data=True)```

Hopefully that made sense, looking forward to your reply! 
Blair

**Comment by lmcinnes** [comment]
Hi, thanks for the feedback.

With regard to Problem 1 the sort answer is yes, the way the model is constructed if you have *enough* data densely packed into a small region it will potentially separate out clusters there simply because they are dense regions separated by *relatively* sparse regions. Potentially you can add a very small amount of noise to your data to perturb it a little if you feel there is some uncertainty in location and don't want to have points directly on top of each other.

Problem 2 is directly due to Problem 1 -- the data points are dense enough in very small clumps that, relatively speaking, anything no falling directly in that particular spot counts as noise. As for tweaking hyperparameters -- you could potentially  get around some of these problems by using a large ``min_cluster_size``. Note that setting ``min_cluster_size`` without also specifying ``min_samples`` will set ``min_samples`` to the ``min_cluster_size`` value, which is probably not desireable for this case. Instead you might want something like

    hdbscan.HDBSCAN(min_samples=100, min_cluster_size=1000, gen_min_span_tree=True, metric='haversine', prediction_data=True)

Furthermore if you are fixing ``min_samples`` and varying min_cluster size you can potentially save yourself some computation time with caching. That is you can do something like

    labels1 = hdbscan.HDBSCAN(min_samples=100, min_cluster_size=1000, memory="cachefile, "metric='haversine').fit_predict(data)
    labels2 = hdbscan.HDBSCAN(min_samples=100, min_cluster_size=2000, memory="cachefile", metric='haversine').fit_predict(data)
    labels3 = hdbscan.HDBSCAN(min_samples=100, min_cluster_size=3000, memory="cachefile", metric='haversine').fit_predict(data)

and the subsequent calls of HDBSCAN will be much faster since it will cache much of the hard computational work that depends only on ``min_samples`` and only recomputed that part of the algorithm that depends on ``min_cluster_size``. This may help you in exploring hyperparameter values.

**Comment by lmcinnes** [comment]
Problem 3 sounds like it may be a bug, but potentially an odd one. I'll try to look into that when I get some time.

**Comment by Blair-Young** [comment]
Hey,
Just as a follow up, your suggestion to introduce noise was perfect! Completely cleared up my problem. The noise made the hyper dense areas much more manageable to work with.

As for Problem 3, I was being an idiot, so don't worry about that.
